www/js/app.js

```diff
@@ -64,34 +65,100 @@ define(['jquery', 'zimArchiveLoader', 'uiUtil', 'settingsStore','abstractFilesys
     appstate['search'] = {
         'prefix': '', // A field to hold the original search string
         'status': '',  // The status of the search: ''|'init'|'interim'|'cancelled'|'complete'
         'type': ''    // The type of the search: 'basic'|'full' (set automatically in search algorithm)
     };
+
+    // A Boolean to store the update status of the PWA version (currently only used with Firefox Extension)
+    appstate['updateNeeded'] = false; // This will be set to true if the Service Worker has an update waiting
     
+    /**
+     * Apply any override parameters that might be in the querystring.
+     * This is used for communication between the PWA and any local code (e.g. Firefox Extension), both ways.
+     * It is also possible for DEV (or user) to launch the app with certain settings, or to unset potentially
+     * problematic settings, by crafting the querystring appropriately.
+     */
+    (function overrideParams() {
+        var rgx = /[?&]([^=]+)=([^&]+)/g;
+        var matches = rgx.exec(window.location.search);
+        while (matches) {
+            if (matches[1] && matches[2]) {
+                var paramKey = decodeURIComponent(matches[1]);
+                var paramVal = decodeURIComponent(matches[2]);
+                if (paramKey !== 'title') {
+                    console.debug('Setting key-pair: ' + paramKey + ':' + paramVal);
+                    // Make values Boolean if 'true'/'false'
+                    paramVal = paramVal === 'true' || (paramVal === 'false' ? false : paramVal);
+                    settingsStore.setItem(paramKey, paramVal, Infinity);
+                    params[paramKey] = paramVal;
+                }
+            }
+            matches = rgx.exec(window.location.search);
+        }
+        // If we are in an extension, send a 'success' message to the extension
+        if (params.extensionURL && ~window.location.href.indexOf(params.PWAServer)) {
+            var message = '?PWA_launch=success';
+            // DEV: To test failure of the PWA, you could pause on next line and set message to '?PWA_launch=fail'
+            // Note that, as a failsafe, the PWA_launch key is set to 'fail' (in the extension) before each PWA launch
+            // so we need to send a 'success' message each time the PWA is launched
+            var frame = document.createElement('iframe');
+            frame.id = 'kiwixComm';
+            frame.style.display = 'none';
+            document.body.appendChild(frame);
+            frame.src = params.extensionURL + '/www/index.html'+ message;
+            // Now remove redundant frame. We cannot use onload, because it doesn't give time for the script to run.
+            setTimeout(function () {
+                var kiwixComm = document.getElementById('kiwixComm');
+                // The only browser which does not support .remove() is IE11, but it will never run this code
+                if (kiwixComm) kiwixComm.remove();
+            }, 3000);
```

Comment:
Instead of creating a dummy iframe, waiting an arbitrary time etc, maybe you could simply use an XHR request?

Reply:
I can try!

Reply:
OK, I seem to remember why I didn't choose XHR. It's because we are making a cross-origin request, and I believed that only iframes are allowed to load cross-origin sites (and we don't need to access the contents of the iframe, merely send a one-way message).  It could be that the message would be received with XHR even if the response is blocked? I could try XHR to find out, if it is conceptually neater than using an iframe?

Reply:
It's worth testing with XHR. It would be much cleaner in my opinion, at least because we can handle errors on this HTTP call.
I also dislike the `setTimeout` here because we have no guarantee on how much time the HTTP call can take : in worst case we might close the iframe before the HTTP call could be sent

Reply:
@mossroy I've tried to use XHR with the following code:
```
            var commReq = new XMLHttpRequest();
            var commUrl = params.referrerExtensionURL + '/www/index.html'+ message;
            commReq.addEventListener('load', function (e) {
                console.debug('XHR should have loaded', e);
            });
            commReq.addEventListener('error', function (err) {
                console.error('There was an error communicating with the extension!', err);
            });
            commReq.open('GET', commUrl);
            commReq.send();
```
Unfortunately, it doesn't work. No error is thrown, and the URL seems to load successfully (I get the console message), but it is merely loading index.html and not running attached JavaScript, so init.js in the extension never loads in order to process the querystring and receive the success message.

This is in fact why I chose iframe in the first place, because an iframe can load, sandboxed, a page and run JS on the page. According to https://stackoverflow.com/questions/3728798/running-javascript-downloaded-with-xmlhttprequest, XMLHttpRequest can't do this: we would have to eval any scripts on the page, but they would run in the wrong context, so it would be useless (as well as dangerous). 

An answer [here](https://stackoverflow.com/questions/41110413/javascript-not-executing-when-loaded-via-xmlhttprequest-but-does-when-loaded-wit) suggests that jQuery can do a hack to run javascript from the remote page, but if you look at the source of the jquery function, you see that it loads scripts into the current document, which is precisely what we don't want to happen (wrong context again).

So I think we're stuck with the iframe method. I took care only to load index.html and init.js (not the whole app), and also it's communicating with a local extension, not over the Web, so 3 full seconds in the timeout should be enough time even on the slowest systems before removing the iframe. However, I can increase it to 5 seconds, or even 10 seconds if you prefer. The user won't notice, it'll be eliminated in the background.